pkg/cmd/drtprod/cli/commands/yamlprocessor.go

```diff
@@ -137,50 +140,97 @@ func setEnv(environment map[string]string, displayOnly bool) error {
 func processTargets(
 	ctx context.Context, targets []target, displayOnly bool, targetNames []string,
 ) error {
+	// targetNameMap is used to check all targets that are provided as user input
 	targetNameMap := make(map[string]struct{})
-	targetMap := make(map[string][]*command)
 	for _, tn := range targetNames {
 		targetNameMap[tn] = struct{}{}
 	}
-	for i := 0; i < len(targets); i++ {
-		targets[i].TargetName = os.ExpandEnv(targets[i].TargetName)
-		t := targets[i]
-		if _, ok := targetNameMap[t.TargetName]; len(targetNames) > 0 && !ok {
-			fmt.Printf("Ignoring execution for target %s\n", t.TargetName)
-			continue
-		}
-		// Generate the commands for each target's steps
-		targetSteps, err := generateCmdsFromSteps(t.TargetName, t.Steps)
-		if err != nil {
-			return err
-		}
-		targetMap[t.TargetName] = targetSteps
+	waitGroupTracker, err := buildTargetCmdsAndRegisterWaitGroups(targets, targetNameMap, targetNames)
+	if err != nil {
+		return err
 	}
 
-	// Use a WaitGroup to execute commands concurrently
+	// Use a WaitGroup to wait for commands executed concurrently
 	wg := sync.WaitGroup{}
-	for targetName, cmds := range targetMap {
+	for _, t := range targets {
+		if _, ok := targetNameMap[t.TargetName]; len(targetNames) > 0 && !ok {
+			// user provided targets do not have this target
+			continue
+		}
 		if displayOnly {
-			displayCommands(targetName, cmds)
+			// if we need to display, there is no wait group to wait for
+			displayCommands(t)
 			continue
 		}
```

Comment:
If I understand this correctly, this check can be moved out of the loop and we can return `nil` from it since it only prints the commands. We could extract the check `len(targetNames) > 0 && !ok` to a function and use it in the loop and the `displayOnly` check.

Reply:
Do you mean that this can be put as a separate loop?

Reply:
Yes, create a helper function
```
func processTarget(t target, targetNameMap map[string]struct{}, targetNames []string) bool {
	_, exists := targetNameMap[t.TargetName]
	return len(targetNames) == 0 || exists
}
```
and use it like this
```
  if displayOnly {
		for _, t := range targets {
			if processTarget(t, targetNameMap, targetNames) {
				displayCommands(t)
			}
		}
		return nil
	}

	// Execute commands concurrently with WaitGroup synchronization
	var wg sync.WaitGroup
	for _, t := range targets {
		if !processTarget(t, targetNameMap, targetNames) {
			continue
		}
		.
		.
```

This way we return early when `displayOnly` is true avoid creating `wg`